File: 12-tic-tac-toe/tic_tac_toe.erl
%% Simple Tic tac toe game written in Erlang
-module(tic_tac_toe).
-author("M.V.Harish Kumar").
-compile([{date, "2023-12-08"}]).
-export([main/0]).

display_board([P1, P2, P3, P4, P5, P6, P7, P8, P9]) ->
    io:format("\t +-------+-------+-------+~n"),
    io:format("\t |       |       |       |~n"),
    io:format("\t |   ~p   |   ~p   |   ~p   |~n", [P1, P2, P3]),
    io:format("\t |       |       |       |~n"),
    io:format("\t +-------+-------+-------+~n"),
    io:format("\t |       |       |       |~n"),
    io:format("\t |   ~p   |   ~p   |   ~p   |~n", [P4, P5, P6]),
    io:format("\t |       |       |       |~n"),
    io:format("\t +-------+-------+-------+~n"),
    io:format("\t |       |       |       |~n"),
    io:format("\t |   ~p   |   ~p   |   ~p   |~n", [P7, P8, P9]),
    io:format("\t |       |       |       |~n"),
    io:format("\t +-------+-------+-------+~n").

victory_for([P1, P2, P3, P4, P5, P6, P7, P8, P9], Sign) ->
    ((P1 =:= Sign) and (P2 =:= Sign) and (P3 =:= Sign)) orelse
    ((P4 =:= Sign) and (P5 =:= Sign) and (P6 =:= Sign)) orelse
    ((P7 =:= Sign) and (P8 =:= Sign) and (P9 =:= Sign)) orelse
    ((P1 =:= Sign) and (P4 =:= Sign) and (P7 =:= Sign)) orelse
    ((P2 =:= Sign) and (P5 =:= Sign) and (P8 =:= Sign)) orelse
    ((P3 =:= Sign) and (P6 =:= Sign) and (P9 =:= Sign)) orelse
    ((P1 =:= Sign) and (P5 =:= Sign) and (P9 =:= Sign)) orelse
    ((P3 =:= Sign) and (P5 =:= Sign) and (P7 =:= Sign)).

free_fields(Board) -> [X || X <- Board, is_integer(X)].

move(Sign, Board, Pos) -> 
    case lists:member(Pos, free_fields(Board)) of
        true -> lists:sublist(Board, Pos-1) ++ [Sign] ++ lists:nthtail(Pos, Board);
        false -> io:format("Invalid Input~n")
    end.

play(user, Board) ->
    UserMoves = free_fields(Board),
    if
        UserMoves == [] ->
            io:format("Draw!~n");

        true ->
            display_board(Board),
            {ok, [Pos]} = io:fread("Enter your move: ", "~d"),
            MovedBoard = move(o, Board, Pos),
            case victory_for(MovedBoard, o) of
                true ->
                    display_board(MovedBoard),
                    io:format("You win the match!~n");
                false -> play(computer, MovedBoard)
            end
    end;


play(computer, Board) ->
    CompMoves = free_fields(Board),
    if
        CompMoves == [] ->
            io:format("Draw!~n");

        true ->
            Pos = lists:nth(rand:uniform(length(CompMoves)), CompMoves),
            MovedBoard = move(x, Board, Pos),
            case victory_for(MovedBoard, x) of
                true ->
                    display_board(MovedBoard),
                    io:format("The Computer wins the match!~n");
                false -> play(user, MovedBoard)
            end
    end.

main() ->
    play(user, [1,2,3,4,x,6,7,8,9]),
    erlang:halt(0).

File: 12-tic-tac-toe/tic_tac_toe.exs
# Simple tic tac toe game written in Elixir
# Written by M.V.Harish Kumar on 09-02-2025

defmodule TicTacToe do
  def play(:user, board) do
    case free_fields(board) do
      [] -> IO.puts "Draw!"
      _ ->
        display_board(board)
        pos = IO.gets("Enter your move: ") |> String.trim |> String.to_integer
        moved_board = move(:o, board, pos)
        if victory_for(moved_board, :o) do
          display_board(moved_board)
          IO.puts "You win the match"
        else
          play(:computer, moved_board)
        end
    end
  end

  def play(:computer, board) do
    comp_moves = free_fields(board)
    case comp_moves do
      [] -> IO.puts "Draw!"
      _ ->
        pos = Enum.random(comp_moves)
        moved_board = move(:x, board, pos)
        if victory_for(moved_board, :x) do
          display_board(moved_board)
          IO.puts "The Computer wins the match!"
        else
          play(:user, moved_board)
        end
    end
  end

  defp display_board([p1, p2, p3, p4, p5, p6, p7, p8, p9]) do
    IO.puts "\t +-------+-------+-------+"
    IO.puts "\t |       |       |       |"
    IO.puts "\t |   #{p1}   |   #{p2}   |   #{p3}   |"
    IO.puts "\t |       |       |       |"
    IO.puts "\t +-------+-------+-------+"
    IO.puts "\t |       |       |       |"
    IO.puts "\t |   #{p4}   |   #{p5}   |   #{p6}   |"
    IO.puts "\t |       |       |       |"
    IO.puts "\t +-------+-------+-------+"
    IO.puts "\t |       |       |       |"
    IO.puts "\t |   #{p7}   |   #{p8}   |   #{p9}   |"
    IO.puts "\t |       |       |       |"
    IO.puts "\t +-------+-------+-------+"
  end

  defp victory_for(board, sign) do
    set = List.duplicate(sign, 3)
    (Enum.slice(board, 0..2) === set) ||
    (Enum.slice(board, 3..5) === set) ||
    (Enum.slice(board, 6..8) === set) ||
    (Enum.slice(board, 0..8//3) === set) ||
    (Enum.slice(board, 1..8//3) === set) ||
    (Enum.slice(board, 2..8//3) === set) ||
    (Enum.slice(board, 0..8//4) === set) ||
    (Enum.slice(board, 2..7//2) === set)
  end

  defp free_fields(board) do
    for x <- board, is_integer(x), do: x
  end

  defp move(sign, board, pos) do
    if pos in free_fields(board) do
      List.update_at(board, pos-1, fn _ -> sign end)
    else
      IO.puts "Invalid Input"
    end
  end
end

TicTacToe.play(:user, List.update_at(Enum.to_list(1..9), 4, fn _ -> :x end))
